### ISSUE TITLE
incompatible with openssl 1.1.0?

### ISSUE THREAD

**Issue by johnnybubonic** [archiecobbs/mtree-port#10]
getting this on compile:

```
misc.o: In function `MD5_File':
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:165: undefined reference to `MD5_Init'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:165: undefined reference to `MD5_Update'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:165: undefined reference to `MD5_Final'
misc.o: In function `SHA1_File':
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:169: undefined reference to `SHA1_Init'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:169: undefined reference to `SHA1_Update'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:169: undefined reference to `SHA1_Final'
misc.o: In function `SHA256_File':
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:170: undefined reference to `SHA256_Init'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:170: undefined reference to `SHA256_Update'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:170: undefined reference to `SHA256_Final'
misc.o: In function `RIPEMD160_File':
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:174: undefined reference to `RIPEMD160_Init'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:174: undefined reference to `RIPEMD160_Update'
/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3/misc.c:174: undefined reference to `RIPEMD160_Final'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:435: mtree] Error 1
make[1]: Leaving directory '/var/tmp/pkgbuild-0/mtree/src/mtree-port-1.0.3'
make: *** [Makefile:338: all] Error 2
```

**Comment by archiecobbs** [comment]
Should be fixed by 0380aa3.

**Comment by Earnestly** [comment]
I don't believe it has been fixed, but right now the build completes without error however no checksums are generated in mtree output.

```
% ls
bar  foo

% mtree -Pick sha256digest -p .
#	   user: (null)
#	machine: teapot
#	   tree: /home/earnest/t
#	   date: Fri Dec 22 13:46:14 2017

# .
/set type=file
.               type=dir
    bar
    foo
..
```

The `configure` script also claims it cannot find any of the `EVP_` functions.

Keep in mind that the `EVP_` functions are defined in `openssl/evp.h` which isn't included in most of the files.

I'm not sure what to do, and try as I might, I just cannot seem to understand autotools.

**Comment by archiecobbs** [comment]
OK try again.. should be (really) fixed now in ccffb526.

**Comment by Earnestly** [comment]
Ah, I've already tried that, but couldn't figure out why it didn't help.  For example `./configure' still returns:

```
[...]
checking openssl/sha.h usability... yes
checking openssl/sha.h presence... yes
checking for openssl/sha.h... yes
checking openssl/ripemd.h usability... yes
checking openssl/ripemd.h presence... yes
checking for openssl/ripemd.h... yes
checking for EVP_md5... no
checking for EVP_sha1... no
checking for EVP_sha256... no
checking for EVP_sha384... no
checking for EVP_sha512... no
checking for EVP_ripemd160... no
```

I've also tried to use `PKG_CHECK_MODULES([openssl], [openssl])` to generate "`openssl_LIBS = -lssl -lcrypto`", and modified `Makefile.am` to use `mtree_LDFLAGS = $(openssl_LIBS)` (sadly there doesn't appear to be a `_LDLIBS`).  And while this adds the nedded flags during the linking phase, it still doesn't seem to help.

But no matter what I try, `mtree` still seems to not create checksums when using `-k sha256digest`.

**Comment by archiecobbs** [comment]
The `./configure` output you quoted doesn't show these lines:
```
checking openssl/evp.h usability... yes
checking openssl/evp.h presence... yes
checking for openssl/evp.h... yes
```
which makes me think the change did not get fully applied.

Did you run `./autogen.sh` after applying the patch? This is required to regenerate the `./configure` script from `autoconf.ac`.

**Comment by Earnestly** [comment]
Sorry, I simply didn't include those lines.  They are there.

In fact, here's the output from the entire build process I have locally:

https://ptpb.pw/k6_i

**Comment by archiecobbs** [comment]
This may be the problem:
```
checking for CRYPTO_thread_id in -lcrypto... no
```
If `libcrypto` is not found then lots of crypto stuff will be missing.

Maybe it's under a different `--prefix` ?

**Comment by Earnestly** [comment]
On my system it's under `/usr/lib/libcrypto.so`, which is found using `pkg-config` without issue:

```
# Note, no -L flags are necessary.
pkg-config --cflags --libs libcrypto
-lcrypto
```

**Comment by archiecobbs** [comment]
It might be we're just looking for an obsolete symbol in libcrypto. Try this patch...
```
--- a/configure.ac
+++ b/configure.ac
@@ -27,7 +27,7 @@ AC_PROG_INSTALL
 AC_PROG_CC
 
 # Check for libcrypto library
-AC_CHECK_LIB(crypto, CRYPTO_thread_id,,)
+AC_CHECK_LIB(crypto, CRYPTO_malloc,,)
 
 # Check for required header files
 AC_HEADER_STDC
```
You'll need to re-run `./autogen.sh` and `./configure` of course.

**Comment by Earnestly** [comment]
That appears to be the fix.  Now the `EVP_` checks appear to succeed;

```
checking for EVP_md5... yes
checking for EVP_sha1... yes
checking for EVP_sha256... yes
checking for EVP_sha384... yes
checking for EVP_sha512... yes
checking for EVP_ripemd160... yes
```

And mtree produces the necessary digest:

```
% ./mtree -ck sha256digest -p scripts
#	   user: (null)
#	machine: teapot
#	   tree: /home/earnest/build/pkgbuilds/mtree-git/src/mtree-port/scripts
#	   date: Sat Dec 23 18:59:21 2017

# .
/set type=file
.               type=dir
    compile     \
                sha256digest=2dd0914838c05c6094e2d2348607f5d8fcdce7dafed37d19344adcc4aaca7a5b
    depcomp     \
                sha256digest=534cc0f4c4bfc8f2c5a674741e46abfcc033776a011c3d5bc48cad7d0394ef91
    install-sh  \
                sha256digest=cba2fc55db32a081d81991172dd7b31ef95476eec7fdd887490f7de832215d4e
    missing     \
                sha256digest=2964c9849ee7437665cd86bc13e2018cd4c7d88ad405730c645a88dc3add6d14
..
```

A surprisingly simply fix in the end, thank you o/

**Comment by archiecobbs** [comment]
Great, thanks for testing. Fixed (again) in e11b065.
